examples/shell/shell_common/cmd_ping.cpp

```diff
@@ -0,0 +1,436 @@
+Messaging::ExchangeManager gExchangeManager;
+SecureSessionMgr gSessionManager;
+Inet::IPAddress gDestAddr;
+
+// Max value for the number of echo requests sent.
+uint32_t gMaxEchoCount = 3;
+
+// The CHIP Echo interval time in milliseconds.
+uint32_t gEchoInterval = 1000;
+
+// The last time a echo request was attempted to be sent.
+uint64_t gLastEchoTime = 0;
+
+// Count of the number of echo requests sent.
+uint64_t gEchoCount = 0;
+
+// Count of the number of echo responses received.
+uint64_t gEchoRespCount = 0;
+
+// True, if the echo client is waiting for an echo response
+// after sending an echo request, false otherwise.
+bool gWaitingForEchoResp = false;
+
+bool gUseTCP = false;
+
+bool gUseCRMP = true;
+
+void ResetGlobalState()
+{
+    gMaxEchoCount       = 3;
+    gEchoInterval       = 1000;
+    gLastEchoTime       = 0;
+    gEchoCount          = 0;
+    gEchoRespCount      = 0;
+    gWaitingForEchoResp = false;
+    gUseTCP             = false;
+    gUseCRMP            = true;
+}
+
+bool EchoIntervalExpired(void)
+{
+    uint64_t now = System::Timer::GetCurrentEpoch();
+
+    return (now >= gLastEchoTime + gEchoInterval);
+}
+
+CHIP_ERROR SendEchoRequest(streamer_t * stream)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Messaging::SendFlags sendFlags;
+    const char kRequestFormat[] = "Echo Message %" PRIu64 "\n";
+    char requestData[(sizeof kRequestFormat) + 20 /* uint64_t decimal digits */];
+    snprintf(requestData, sizeof requestData, kRequestFormat, gEchoCount);
+    System::PacketBufferHandle payloadBuf = MessagePacketBuffer::NewWithData(requestData, strlen(requestData));
+
+    if (gUseCRMP)
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNone);
+    }
+    else
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNoAutoRequestAck);
+    }
+
+    if (payloadBuf.IsNull())
+    {
+        streamer_printf(stream, "Unable to allocate packet buffer\n");
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    gLastEchoTime = System::Timer::GetCurrentEpoch();
+
+    streamer_printf(stream, "\nSend echo request message to Node: %" PRIu64 "\n", kTestDeviceNodeId);
+
+    err = gEchoClient.SendEchoRequest(std::move(payloadBuf), sendFlags);
+
+    if (err == CHIP_NO_ERROR)
+    {
+        gWaitingForEchoResp = true;
+        gEchoCount++;
+    }
+    else
+    {
+        streamer_printf(stream, "Send echo request failed, err: %s\n", ErrorStr(err));
+    }
+
+    return err;
+}
+
+CHIP_ERROR EstablishSecureSession(streamer_t * stream)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Optional<Transport::PeerAddress> peerAddr;
+    SecurePairingUsingTestSecret * testSecurePairingSecret = chip::Platform::New<SecurePairingUsingTestSecret>();
+    VerifyOrExit(testSecurePairingSecret != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    if (gUseTCP)
+    {
+        peerAddr = Optional<Transport::PeerAddress>::Value(Transport::PeerAddress::TCP(gDestAddr, CHIP_PORT));
```

Comment:
Should the remote peer port also be configurable?

Reply:
A CHIP node should be able to expose its CHIP service(s) at any port(s) it chooses.  Having a well-known default is fine.  Constraining to the default is not.

Reply:
Acked, add option '-p' to specify the which port the Echo message is sent to, use default CHIP_PORT if not specified.